datafusion/physical-plan/src/repartition/mod.rs

```diff
@@ -1540,7 +1550,46 @@ impl Stream for PerPartitionStream {
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Option<Self::Item>> {
-        let poll = self.poll_next_inner(cx);
+        let poll = match self.batch_coalescer.take() {
+            Some(mut coalescer) => {
+                let cloned_time = self.baseline_metrics.elapsed_compute().clone();
+                let mut completed = false;
+                let poll;
+                loop {
+                    if let Some(batch) = coalescer.next_completed_batch() {
+                        poll = Poll::Ready(Some(Ok(batch)));
+                        break;
+                    }
+                    if completed {
+                        poll = Poll::Ready(None);
+                        break;
+                    }
+                    let inner_poll = self.poll_next_inner(cx);
+                    let _timer = cloned_time.timer();
+
+                    match inner_poll {
+                        Poll::Pending => {
+                            poll = Poll::Pending;
+                            break;
+                        }
+                        Poll::Ready(None) => {
+                            completed = true;
+                            coalescer.finish()?;
```

Comment:
```suggestion
                            if let Err(e) = coalescer.finish() {
                                self.batch_coalescer = Some(coalescer);
                                return Err(e);
                            }
```
Otherwise in case of an error the `self.batch_coalescer` won't be restored.

Reply:
I think in this case, the loop would continue and then `break` on line  1561 or 1565. Then line 1588 is reached, which restores the `self.batch_coalescer`, but please let me know if that does not make sense.

Reply:
ah thanks for catching this and for the explanation, updated in revision.